src/persist/src/sqlite.rs

```diff
@@ -0,0 +1,167 @@
+                 WHERE shard = $shard ORDER BY sequence_number DESC LIMIT 1",
+        )?;
+        stmt.query_row(named_params! {"$shard": self.shard}, |row| {
+            let sequence_number = row.get("sequence_number")?;
+            let data: Vec<_> = row.get("data")?;
+            Ok(VersionedData {
+                seqno: SeqNo(sequence_number),
+                data,
+            })
+        })
+        .optional()
+        .map_err(|e| e.into())
+    }
+
+    async fn compare_and_set(
+        &mut self,
+        deadline: Instant,
+        expected: Option<SeqNo>,
+        new: VersionedData,
+    ) -> Result<Result<(), Option<VersionedData>>, Error> {
+        if let Some(expected) = expected {
+            if new.seqno <= expected {
+                return Err(Error::from(
+                        format!("new seqno must be strictly greater than expected. Got new: {:?} expected: {:?}",
+                                 new.seqno, expected)));
+            }
+        }
+
+        let result = if let Some(expected) = expected {
+            let conn = self.conn.lock().await;
+
+            // Only insert the new row if:
+            // - sequence number expected is already present
+            // - expected corresponds to the most recent sequence number
+            //   i.e. there is no other sequence number > expected already
+            //   present.
+            let mut stmt = conn
+                .prepare_cached(
+                    "INSERT INTO consensus SELECT $shard, $sequence_number, $data WHERE
+                     EXISTS (
+                        SELECT * FROM consensus WHERE shard = $shard AND sequence_number = $expected
+                     ) 
+                     AND NOT EXISTS (
+                         SELECT * FROM consensus WHERE shard = $shard AND sequence_number > $expected
+                     )
+                     ON CONFLICT DO NOTHING",
+                )?;
+
+            stmt.execute(named_params! {
+                "$shard": self.shard,
+                "$sequence_number": new.seqno.0,
+                "$data": new.data,
+                "$expected": expected.0,
+            })?
+        } else {
+            let conn = self.conn.lock().await;
+
+            // Insert the new row as long as no other row exists for the same shard.
+            let mut stmt = conn.prepare_cached(
+                "INSERT INTO consensus SELECT $shard, $sequence_number, $data WHERE
+                     NOT EXISTS (
+                         SELECT * FROM consensus WHERE shard = $shard
+                     ) 
+                     ON CONFLICT DO NOTHING",
+            )?;
+            stmt.execute(named_params! {
+                "$shard": self.shard,
+                "$sequence_number": new.seqno.0,
+                "$data": new.data,
+            })?
+        };
+
+        if result == 1 {
+            Ok(Ok(()))
+        } else {
+            let current = self.head(deadline).await?;
```

Comment:
add a (very quick) note please about why the race condition of having this in a follow txn is safe

Reply:
done!